File: packages/pinjected-linter/src/pinjected_linter/rules/pinj011_iproxy_annotations.py
"""PINJ011: IProxy type annotations rule."""

import ast

from ..models import RuleContext, Severity
from ..utils.ast_utils import has_decorator
from .base import ASTRuleBase, BaseNodeVisitor


class PINJ011Visitor(BaseNodeVisitor):
    """Visitor for checking IProxy type annotations."""

    def __init__(self, rule, context):
        super().__init__(rule, context)
        self.has_iproxy_import = False
        self.iproxy_alias = None

    def visit_ImportFrom(self, node):
        """Track IProxy imports."""
        if node.module == "pinjected":
            for alias in node.names:
                if alias.name == "IProxy":
                    self.has_iproxy_import = True
                    self.iproxy_alias = alias.asname or "IProxy"
        self.generic_visit(node)

    def visit_FunctionDef(self, node):
        """Check function definitions."""
        self._check_function(node)

    def visit_AsyncFunctionDef(self, node):
        """Check async function definitions."""
        self._check_function(node)

    def _check_function(self, node):
        """Check function for IProxy usage patterns."""
        # Check if this is an @injected function
        if has_decorator(node, "injected"):
            self._check_injected_function(node)
        # Check if this is an @instance function
        elif has_decorator(node, "instance"):
            self._check_instance_function(node)

        self.generic_visit(node)

    def _check_injected_function(self, node):
        """Check @injected functions for IProxy usage."""
        # In Python AST, positional-only args (before /) are in posonlyargs
        # Regular args (after /) are in args

        if not node.args.posonlyargs:
            # No positional-only args means no dependencies (no slash or slash at beginning)
            return

        # Check injected parameters (positional-only args before slash)
        for arg in node.args.posonlyargs:
            annotation = arg.annotation

            # Check if the parameter looks like a dependency that should use IProxy
            if self._should_use_iproxy(arg.arg, annotation):
                self.add_violation(
                    arg,
                    f"Parameter '{arg.arg}' in @injected function '{node.name}' should use IProxy[T] annotation. "
                    f"Dependencies that are injected should be typed with IProxy for proper type checking.",
                    suggestion=f"Change type annotation to IProxy[{self._get_inner_type(annotation)}]",
                )

    def _check_instance_function(self, node):
        """Check @instance function return types."""
        # @instance functions that provide entry points should return IProxy
        return_annotation = node.returns

        if (
            return_annotation
            and self._is_service_type(return_annotation)
            and not self._is_iproxy_annotation(return_annotation)
        ):
            self.add_violation(
                node,
                f"@instance function '{node.name}' returns a service type but doesn't use IProxy. "
                f"Entry point services should return IProxy[T] for proper dependency tracking.",
                suggestion=f"Change return type to IProxy[{self._get_annotation_string(return_annotation)}]",
            )

    def _should_use_iproxy(self, param_name: str, annotation) -> bool:
        """Check if a parameter should use IProxy based on naming and type."""
        # Skip if already using IProxy
        if annotation and self._is_iproxy_annotation(annotation):
            return False

        # Only suggest IProxy if there's a type annotation that looks like a service
        if not annotation:
            return False

        # Check type annotation
        annotation_str = self._get_annotation_string(annotation)

        # Service type patterns in annotations
        service_patterns = [
            "Service",
            "Client",
            "Repository",
            "Manager",
            "Factory",
            "Provider",
            "Handler",
            "Processor",
            "Validator",
            "Logger",
            "Database",
            "Cache",
            "Gateway",
            "Controller",
        ]

        # Check if the annotation looks like a service type
        return any(pattern in annotation_str for pattern in service_patterns)

    def _is_service_type(self, annotation) -> bool:
        """Check if the type looks like a service/component."""
        if not annotation:
            return False

        annotation_str = self._get_annotation_string(annotation)
        service_patterns = [
            "Service",
            "Client",
            "Repository",
            "Manager",
            "Handler",
            "Controller",
            "Gateway",
            "Adapter",
            "Provider",
        ]

        return any(pattern in annotation_str for pattern in service_patterns)

    def _is_iproxy_annotation(self, annotation) -> bool:
        """Check if the annotation is IProxy[T] or just IProxy."""
        if isinstance(annotation, ast.Name):
            # Check for bare IProxy
            return annotation.id == (self.iproxy_alias or "IProxy")
        elif isinstance(annotation, ast.Subscript) and isinstance(
            annotation.value, ast.Name
        ):
            return annotation.value.id == (self.iproxy_alias or "IProxy")
        return False

    def _get_annotation_string(self, annotation) -> str:
        """Get string representation of annotation."""
        if isinstance(annotation, ast.Name):
            return annotation.id
        elif isinstance(annotation, ast.Subscript):
            return self._get_annotation_string(annotation.value)
        elif isinstance(annotation, ast.Attribute):
            value = self._get_annotation_string(annotation.value)
            return f"{value}.{annotation.attr}"
        return "Any"

    def _get_inner_type(self, annotation) -> str:
        """Get the inner type for IProxy suggestion."""
        if annotation:
            return self._get_annotation_string(annotation)
        return "Any"


class PINJ011IProxyAnnotations(ASTRuleBase):
    """Rule for checking IProxy type annotations.

    IProxy[T] should be used for:
    1. Type annotations of injected dependencies in @injected functions
    2. Return types of @instance functions that provide entry points
    3. References to dependencies that are resolved through injection

    This helps with:
    - Type safety in dependency injection
    - Clear distinction between injected and runtime values
    - Better IDE support and type checking
    """

    rule_id = "PINJ011"
    name = "IProxy type annotations"
    description = "Dependencies should use IProxy[T] type annotations"
    severity = Severity.WARNING
    category = "typing"
    auto_fixable = False  # Type changes require careful consideration

    def get_visitor(self, context: RuleContext) -> ast.NodeVisitor:
        """Get the AST visitor for this rule."""
        return PINJ011Visitor(self, context)

File: packages/pinjected-linter/rust-poc/src/rules/pinj011_iproxy_annotations.rs
//! PINJ011: IProxy type annotations
//! 
//! IProxy[T] should be used for:
//! 1. Return types of @instance functions when they are entry points
//! 
//! IProxy[T] should NOT be used for:
//! 1. Injected dependencies (parameters before / in @injected functions)
//! 2. Regular function parameters or return types

use rustpython_ast::{Mod, Stmt, Expr};
use crate::models::{Violation, RuleContext, Severity};
use crate::rules::base::LintRule;
use crate::utils::pinjected_patterns::{has_instance_decorator, has_instance_decorator_async};

pub struct IProxyAnnotationsRule {
    /// Track if IProxy is imported and its alias
    has_iproxy_import: bool,
    iproxy_alias: Option<String>,
}

impl IProxyAnnotationsRule {
    pub fn new() -> Self {
        Self {
            has_iproxy_import: false,
            iproxy_alias: None,
        }
    }
    
    /// Track IProxy imports
    fn track_imports(&mut self, ast: &Mod) {
        match ast {
            Mod::Module(module) => {
                for stmt in &module.body {
                    if let Stmt::ImportFrom(import) = stmt {
                        if let Some(module) = &import.module {
                            if module.as_str() == "pinjected" {
                                for alias in &import.names {
                                    if alias.name.as_str() == "IProxy" {
                                        self.has_iproxy_import = true;
                                        self.iproxy_alias = alias.asname.as_ref().map(|s| s.to_string());
                                    }
                                }
                            }
                        }
                    }
                }
            }
            _ => {}
        }
    }
    
    /// Check if the annotation is IProxy[T] or just IProxy
    fn is_iproxy_annotation(&self, annotation: &Expr) -> bool {
        match annotation {
            Expr::Name(name) => {
                // Check for bare IProxy
                let iproxy_name = self.iproxy_alias.as_deref().unwrap_or("IProxy");
                name.id.as_str() == iproxy_name
            }
            Expr::Subscript(subscript) => {
                // Check for IProxy[T]
                if let Expr::Name(name) = &*subscript.value {
                    let iproxy_name = self.iproxy_alias.as_deref().unwrap_or("IProxy");
                    name.id.as_str() == iproxy_name
                } else {
                    false
                }
            }
            _ => false
        }
    }
    
    /// Check if the type looks like a service/component
    fn is_service_type(&self, annotation: &Expr) -> bool {
        let annotation_str = self.get_annotation_string(annotation);
        
        let service_patterns = [
            "Service", "Client", "Repository", "Manager",
            "Handler", "Controller", "Gateway", "Adapter", "Provider",
        ];
        
        service_patterns.iter().any(|pattern| annotation_str.contains(pattern))
    }
    
    /// Get string representation of annotation
    fn get_annotation_string(&self, annotation: &Expr) -> String {
        match annotation {
            Expr::Name(name) => name.id.to_string(),
            Expr::Subscript(subscript) => {
                // For subscripted types like List[int], get the base type
                self.get_annotation_string(&subscript.value)
            }
            Expr::Attribute(attr) => {
                let value = self.get_annotation_string(&attr.value);
                format!("{}.{}", value, attr.attr)
            }
            _ => "Any".to_string()
        }
    }
    
    /// Check @instance function return types
    fn check_instance_function(&self, func: &rustpython_ast::StmtFunctionDef, file_path: &str, violations: &mut Vec<Violation>) {
        if let Some(return_annotation) = &func.returns {
            // Check if it's a service type that should use IProxy
            if self.is_service_type(return_annotation) && !self.is_iproxy_annotation(return_annotation) {
                let annotation_str = self.get_annotation_string(return_annotation);
                
                violations.push(Violation {
                    rule_id: "PINJ011".to_string(),
                    message: format!(
                        "@instance function '{}' returns a service type but doesn't use IProxy. \
                        Entry point services should return IProxy[T] for proper dependency tracking. \
                        Change return type to IProxy[{}]",
                        func.name,
                        annotation_str
                    ),
                    offset: func.range.start().to_usize(),
                    file_path: file_path.to_string(),
                    severity: Severity::Warning,
                });
            }
        }
    }
    
    /// Check async @instance function return types
    fn check_instance_function_async(&self, func: &rustpython_ast::StmtAsyncFunctionDef, file_path: &str, violations: &mut Vec<Violation>) {
        if let Some(return_annotation) = &func.returns {
            // Check if it's a service type that should use IProxy
            if self.is_service_type(return_annotation) && !self.is_iproxy_annotation(return_annotation) {
                let annotation_str = self.get_annotation_string(return_annotation);
                
                violations.push(Violation {
                    rule_id: "PINJ011".to_string(),
                    message: format!(
                        "@instance function '{}' returns a service type but doesn't use IProxy. \
                        Entry point services should return IProxy[T] for proper dependency tracking. \
                        Change return type to IProxy[{}]",
                        func.name,
                        annotation_str
                    ),
                    offset: func.range.start().to_usize(),
                    file_path: file_path.to_string(),
                    severity: Severity::Warning,
                });
            }
        }
    }
    
    /// Check statements
    fn check_stmt(&self, stmt: &Stmt, file_path: &str, violations: &mut Vec<Violation>) {
        match stmt {
            Stmt::FunctionDef(func) => {
                // Only check @instance functions
                if has_instance_decorator(func) {
                    self.check_instance_function(func, file_path, violations);
                }
                // Check nested functions
                for stmt in &func.body {
                    self.check_stmt(stmt, file_path, violations);
                }
            }
            Stmt::AsyncFunctionDef(func) => {
                // Only check @instance functions
                if has_instance_decorator_async(func) {
                    self.check_instance_function_async(func, file_path, violations);
                }
                // Check nested functions
                for stmt in &func.body {
                    self.check_stmt(stmt, file_path, violations);
                }
            }
            Stmt::ClassDef(class) => {
                // Check methods in classes
                for stmt in &class.body {
                    self.check_stmt(stmt, file_path, violations);
                }
            }
            _ => {}
        }
    }
}

impl LintRule for IProxyAnnotationsRule {
    fn rule_id(&self) -> &str {
        "PINJ011"
    }
    
    fn description(&self) -> &str {
        "Dependencies should use IProxy[T] type annotations"
    }

    fn check(&self, context: &RuleContext) -> Vec<Violation> {
        let mut violations = Vec::new();
        
        // Create a mutable instance for stateful tracking
        let mut checker = IProxyAnnotationsRule::new();
        
        // First track imports
        checker.track_imports(context.ast);
        
        // Then check the current statement
        checker.check_stmt(context.stmt, context.file_path, &mut violations);
        
        violations
    }
}